Update New Android API Docs (#180604)
flutter/flutter: docs/platforms/android/New-Android-version.md

## Objective

> Provides a list of areas to consider and examples of former work for how to update Flutter to support a new version of the Android API. This generally happens every fall and Flutter developers expect to build against the latest versions quickly after they are available.

Reviewer: I think we should say fall and spring now that Android releases biannually.

Author: done

### Overview

Whenever Android releases a new API version, we have to ensure Flutter apps on Android continue to successfully build on that new version. We achieve this by addressing new Android API behavioral and breaking changes when necessary and updating our infrastructure to test against the new Android API.
Addressing breaking changes will differentiate every year, while updating the Flutter on Android CI will remain similar from year to year.

Below is a guide to ensuring Flutter apps on Android work on the new API. It is recommended that you follow the directions in order. If something is blocked, move on to the next step.

> Please update this doc whenever there is a new Android API.

Author: Updated it to this. It's a little more specific. done
```suggestion
Please maintain this doc to reflect the latest processes, issues, and PRs.
```

#### Umbrella Issue for Updating to the new API

> Updating from API 35 to API 36: <https://github.com/flutter/flutter/issues/163071>

Reviewer: Can you phrase this like a suggestion?

```suggestion
#### Create an Umbrella Issue for Updating to the new API

An example: updating from API 35 to API 36: <https://github.com/flutter/flutter/issues/163071>
```

Author: done

#### New Android features

> New Android features can cause breaking or behavioral changes to Flutter, and they can require a broad spectrum of work. The Flutter Android team investigates these new Android API features and determines which are no-ops and which require work. We are generally aware of breaking changes and schedule work ahead of time.

Reviewer: Is this a step? If so maybe also phrase this as a step:

```suggestion
#### Investigate new Android features

New Android features can cause breaking or behavioral changes to Flutter, and they can require a broad spectrum of work. The Flutter Android team should investigate these new Android API features and determine which are no-ops and which require work. We are generally aware of breaking changes and schedule work ahead of time.
```

Author: done

#### Bump compile and target SDK versions in samples

Samples, especially add to app samples, represent apps that mirror the first types of users we see adopt new Android APIs.

Example PR: https://github.com/flutter/samples/pull/2368

---

### Update Robolectric

Robolectric is a dependency that allows us to write unit tests that run on a local development machine against the Android API surface without being on an Android device.

1. Check online for the Robolectric release notes. It is possible that a Robolectric version supporting the new Android API has not been released yet. You are blocked until that new version is released.

2. Find all usages of Robolectric in `flutter/flutter` (including the engine) and `flutter/packages` and update them.

- Example: <https://github.com/flutter/flutter/issues/177674>.

---

> ### Updating CI

Reviewer: I prefer to leave this as "update" since these are steps and to maintain consistency with the other titles:

```suggestion
### Update CI
```

Author: done

- Example PR: <https://github.com/flutter/flutter/pull/175365>.
- Example PR: <https://github.com/flutter/flutter/pull/179963>.

---

### Update SDK and Dependency Versioning Support

New versions of Android have new dependency version minimums. We must update the Flutter apps we test against. These are existing Flutter apps that test against realistic use cases of Flutter on Android, such as Flutter plugin examples and Flutter integration tests. We must also update the Flutter templates' dependency versions, which will affect newly created Flutter on Android apps. For best practices, we should first update all Flutter Android apps we test against then the `ci.yaml`, as these steps won’t directly affect users. After that, we will update the Flutter templates, which will directly affect users.

#### Update `android_sdk` in `ci.yaml`

Update the `android_sdk` version in `flutter/flutter` and `flutter/packages` ci.yaml to test against the new android sdk. If Flutter must test against a new version of java, you must also update the ci.yaml.

- Example PRs: <https://github.com/flutter/flutter/pull/166464> (relevant changes are only in the `ci.yaml` file)

> - Example PRs: <https://github.com/flutter/packages/pull/9414> (change only the `android_sdk` version)

Reviewer: nit for clarity:

```suggestion
- Example PR where relevant changes are only in the `ci.yaml` file: <https://github.com/flutter/flutter/pull/166464>
- Example PR where change only the `android_sdk` version: <https://github.com/flutter/packages/pull/9414>
```

Author: done

- `shell/platform/android/AndroidManifest.xml`: Bump `android:targetSdkVersion=XX` to the latest version.
- `testing/android/native_activity/native_activity.gni`: Bump the reference to `build-tools/XX` in `android_buildtools` to the latest `build-tools` version and the reference to `android-XX` in `android_jar` to the latest version.

This list may become outdated, so be sure to change any references to the old SDK version to the latest version in `build.gradle` files across the repo.

- Example PR: <https://github.com/flutter/flutter/pull/166796>.

#### Update all existing Flutter on Android apps in our repos

1. Reference the new Android API release notes to see the required dependency version minimums.

2. Update the build dependencies accordingly (new API and dependencies).

* `flutter/flutter` example PR: <https://github.com/flutter/flutter/pull/176858>.
* `flutter/packages` example PR: <https://github.com/flutter/packages/pull/9241>.

#### Update Versions in the Flutter Android Template

1. Reference the new Android API release notes to see the required dependency version minimums.

> 2. Update the Flutter templates to use the new Android API and new dependency versioning minimums. Note: Do not update the template `targetSdk` in this change because additional infra changes may be necessary. Please update the `targetSdk` in a different PR.

Reviewer: I think it would be clearer to reference the targetSdk update step here/create a separate step for it.
